File: grammarinator/generate.py
# Copyright (c) 2017-2025 Renata Hodovan, Akos Kiss.
#
# Licensed under the BSD 3-Clause License
# <LICENSE.rst or https://opensource.org/licenses/BSD-3-Clause>.
# This file may not be copied, modified, or distributed except
# according to those terms.

import json
import os
import random

from argparse import ArgumentParser, SUPPRESS
from functools import partial
from itertools import count
from math import inf
from multiprocessing import Manager, Pool
from os.path import abspath, exists, join

import enlighten

from inators.arg import add_log_level_argument, add_sys_path_argument, add_sys_recursion_limit_argument, add_version_argument, process_log_level_argument, process_sys_path_argument, process_sys_recursion_limit_argument
from inators.imp import import_object

from .cli import add_encoding_argument, add_encoding_errors_argument, add_tree_format_argument, add_jobs_argument, import_list, init_logging, logger, process_tree_format_argument
from .pkgdata import __version__
from .runtime import RuleSize
from .tool import DefaultGeneratorFactory, FilePopulation, GeneratorTool


def int_or_inf(value):
    return inf if value == 'inf' else int(value)


def process_args(args):
    args.generator = import_object(args.generator)
    args.model = import_object(args.model)
    args.listener = import_list(args.listener)
    args.transformer = import_list(args.transformer)
    args.serializer = import_object(args.serializer) if args.serializer else None

    if args.weights:
        if not exists(args.weights):
            raise ValueError('Custom weights should point to an existing JSON file.')

        with open(args.weights, 'r') as f:
            weights = {}
            for rule, alts in json.load(f).items():
                for alternation_idx, alternatives in alts.items():
                    for alternative_idx, w in alternatives.items():
                        weights[(rule, int(alternation_idx), int(alternative_idx))] = w
            args.weights = weights

    if args.population:
        args.population = abspath(args.population)


def generator_tool_helper(args, lock=None):
    return GeneratorTool(generator_factory=DefaultGeneratorFactory(args.generator,
                                                                   model_class=args.model,
                                                                   weights=args.weights,
                                                                   listener_classes=args.listener),
                         rule=args.rule, out_format=args.out, lock=lock,
                         limit=RuleSize(depth=args.max_depth, tokens=args.max_tokens),
                         population=FilePopulation(args.population, args.tree_extension, args.tree_codec) if args.population else None, keep_trees=args.keep_trees,
                         generate=args.generate, mutate=args.mutate, recombine=args.recombine, unrestricted=args.unrestricted,
                         allowlist=args.allowlist, blocklist=args.blocklist,
                         transformers=args.transformer, serializer=args.serializer,
                         memo_size=args.memo_size, unique_attempts=args.unique_attempts,
                         cleanup=False, encoding=args.encoding, errors=args.encoding_errors, dry_run=args.dry_run)


def create_test(generator_tool, index, *, seed):
    if seed:
        random.seed(seed + index)
    return generator_tool.create_test(index)


def execute():
    parser = ArgumentParser(description='Grammarinator: Generate', epilog="""
        The tool acts as a default execution harness for generators
        created by Grammarinator:Processor.
        """)
    # Settings for generating from grammar.
    parser.add_argument('generator', metavar='NAME',
                        help='reference to the generator created by grammarinator-process (in package.module.class format).')
    parser.add_argument('-r', '--rule', metavar='NAME',
                        help='name of the rule to start generation from (default: the parser rule set by grammarinator-process).')
    parser.add_argument('-m', '--model', metavar='NAME', default='grammarinator.runtime.DefaultModel',
                        help='reference to the decision model (in package.module.class format) (default: %(default)s).')
    parser.add_argument('-l', '--listener', metavar='NAME', action='append', default=[],
                        help='reference to a listener (in package.module.class format).')
    parser.add_argument('-t', '--transformer', metavar='NAME', action='append', default=[],
                        help='reference to a transformer (in package.module.function format) to postprocess the generated tree '
                             '(the result of these transformers will be saved into the serialized tree, e.g., variable matching).')
    parser.add_argument('-s', '--serializer', metavar='NAME',
                        help='reference to a seralizer (in package.module.function format) that takes a tree and produces a string from it.')
    parser.add_argument('-d', '--max-depth', default=RuleSize.max.depth, type=int, metavar='NUM',
                        help='maximum recursion depth during generation (default: %(default)f).')
    parser.add_argument('--max-tokens', default=RuleSize.max.tokens, type=int, metavar='NUM',
                        help='maximum token number during generation (default: %(default)f).')
    parser.add_argument('-w', '--weights', metavar='FILE',
                        help='JSON file defining custom weights for alternatives.')

    # Evolutionary settings.
    parser.add_argument('--population', metavar='DIR',
                        help='directory of grammarinator tree pool.')
    parser.add_argument('--no-generate', dest='generate', default=True, action='store_false',
                        help='disable test generation from grammar.')
    parser.add_argument('--no-mutate', dest='mutate', default=True, action='store_false',
                        help='disable test generation by mutation (disabled by default if no population is given).')
    parser.add_argument('--no-recombine', dest='recombine', default=True, action='store_false',
                        help='disable test generation by recombination (disabled by default if no population is given).')
    parser.add_argument('--no-grammar-violations', dest='unrestricted', default=True, action='store_false',
                        help='disable applying grammar-violating mutators (enabled by default)')
    parser.add_argument('--allowlist', metavar='LIST', type=lambda s: [x.strip() for x in s.split(',') if x.strip()],
                        help='comma-separated list of mutators to allow (by default, all mutators are allowed).')
    parser.add_argument('--blocklist', metavar='LIST', type=lambda s: [x.strip() for x in s.split(',') if x.strip()],
                        help='comma-separated list of mutators to block (by default, no mutators are blocked).')
    parser.add_argument('--keep-trees', default=False, action='store_true',
                        help='keep generated tests to participate in further mutations or recombinations (only if population is given).')
    add_tree_format_argument(parser)

    # Auxiliary settings.
    parser.add_argument('-o', '--out', metavar='FILE', default=join(os.getcwd(), 'tests', 'test_%d'),
                        help='output file name pattern (default: %(default)s).')
    parser.add_argument('--stdout', dest='out', action='store_const', const='', default=SUPPRESS,
                        help='print test cases to stdout (alias for --out=%(const)r)')
    parser.add_argument('-n', default=1, type=int_or_inf, metavar='NUM',
                        help='number of tests to generate, \'inf\' for continuous generation (default: %(default)s).')
    parser.add_argument('--memo-size', default=0, type=int, metavar='NUM',
                        help='memoize the last NUM unique tests; if a memoized test case is generated again, it is discarded and generation of a unique test case is retried (default: %(default)s).')
    parser.add_argument('--unique-attempts', default=2, type=int, metavar='NUM',
                        help='limit on how many times to try to generate a unique (i.e., non-memoized) test case; no effect if --memo-size=0 (default: %(default)s).')
    parser.add_argument('--random-seed', type=int, metavar='NUM',
                        help='initialize random number generator with fixed seed (not set by default).')
    parser.add_argument('--dry-run', default=False, action='store_true',
                        help='generate tests without writing them to file or printing to stdout (do not keep generated tests in population either)')
    add_encoding_argument(parser, help='output file encoding (default: %(default)s).')
    add_encoding_errors_argument(parser)
    add_jobs_argument(parser)
    add_sys_path_argument(parser)
    add_sys_recursion_limit_argument(parser)
    add_log_level_argument(parser, short_alias=())
    add_version_argument(parser, version=__version__)
    args = parser.parse_args()

    init_logging()
    process_log_level_argument(args, logger)
    process_sys_path_argument(args)
    process_sys_recursion_limit_argument(args)
    process_tree_format_argument(args)
    try:
        process_args(args)
    except ValueError as e:
        parser.error(e)

    with enlighten.get_manager(enabled=bool(args.out) and args.n != inf) as progress_manager:
        progress_bar = progress_manager.counter(total=args.n, desc='Generating', unit='file')

        if args.jobs > 1:
            with Manager() as manager:
                with generator_tool_helper(args, lock=manager.Lock()) as generator_tool:
                    parallel_create_test = partial(create_test, generator_tool, seed=args.random_seed)
                    with Pool(args.jobs) as pool:
                        for _ in pool.imap_unordered(parallel_create_test, count(0) if args.n == inf else range(args.n)):
                            progress_bar.update()
        else:
            with generator_tool_helper(args) as generator_tool:
                for i in count(0) if args.n == inf else range(args.n):
                    create_test(generator_tool, i, seed=args.random_seed)
                    progress_bar.update()


if __name__ == '__main__':
    execute()

File: grammarinator-cxx/tools/generate.cpp
// Copyright (c) 2025 Renata Hodovan, Akos Kiss.
//
// Licensed under the BSD 3-Clause License
// <LICENSE.rst or https://opensource.org/licenses/BSD-3-Clause>.
// This file may not be copied, modified, or distributed except
// according to those terms.

#include <grammarinator/runtime.hpp>
#include <grammarinator/tool.hpp>
#include <grammarinator/util/print.hpp>
#include <grammarinator/util/random.hpp>

#include <cxxopts.hpp>

#include <filesystem>
#include <map>
#include <random>
#include <string>
#include <tuple>

#include "grammarinator/config.hpp"

#ifndef GRAMMARINATOR_GENERATOR
#error "GRAMMARINATOR_GENERATOR must be defined"
#endif

using namespace grammarinator::runtime;
using namespace grammarinator::tool;
using namespace grammarinator::util;

template<class T>
TreeCodec* treecodec_factory() { return new T(); }

static const std::map<std::string, std::tuple<std::string, TreeCodec*(*)()>> tree_formats = {
  {"flatbuffers", {"grtf", treecodec_factory<FlatBuffersTreeCodec>}},
  {"json", {"grtj", treecodec_factory<JsonTreeCodec>}},
};

// Trim from both ends (in place)
inline void trim(std::string& s) {
    auto not_space = std::not_fn(static_cast<int(*)(int)>(std::isspace));

    s.erase(s.begin(), std::find_if(s.begin(), s.end(), not_space));
    s.erase(std::find_if(s.rbegin(), s.rend(), not_space).base(), s.end());
}

int main(int argc, char **argv) {
  std::string tree_format_choices;
  bool first_format = true;
  for (const auto& tree_format : tree_formats) {
    if (!first_format) {
      tree_format_choices += ", ";
    }
    tree_format_choices += tree_format.first;
    first_format = false;
  }

  try {
    cxxopts::Options options(argv[0], "Grammarinator: Generate (with C++)");
    options.add_options()
      ("r,rule",
       "name of the rule to start generation from (default: the parser rule set by grammarinator-process)",
       cxxopts::value<std::string>(),
       "NAME")
      ("d,max-depth",
       "maximum recursion depth during generation",
       cxxopts::value<int>()->default_value(std::to_string(RuleSize::max().depth)),
       "NUM")
      ("max-tokens",
       "maximum token number during generation",
       cxxopts::value<int>()->default_value(std::to_string(RuleSize::max().tokens)),
       "NUM")
      ("weights",
       "JSON file defining custom weights for alternatives",
       cxxopts::value<std::string>(),
       "FILE")
      ("p,population",
       "directory of grammarinator tree pool",
       cxxopts::value<std::string>(),
       "DIR")
      ("no-generate",
       "disable test generation from grammar",
       cxxopts::value<bool>()->default_value("false"))
      ("no-mutate",
       "disable test generation by mutation (disabled by default if no population is given)",
       cxxopts::value<bool>()->default_value("false"))
      ("no-recombine",
       "disable test generation by recombination (disabled by default if no population is given)",
       cxxopts::value<bool>()->default_value("false"))
      ("no-grammar-violations",
       "disable applying grammar-violating mutators (enabled by default)",
       cxxopts::value<bool>()->default_value("false"))
      ("allowlist",
       "list of enabled test creators",
       cxxopts::value<std::vector<std::string>>())
      ("blocklist",
       "list of disabled test creators",
       cxxopts::value<std::vector<std::string>>())
      ("keep-trees",
       "keep generated tests to participate in further mutations or recombinations (only if population is given)",
       cxxopts::value<bool>()->default_value("false"))
      ("tree-format",
       "format of the serialized trees (choices: " + tree_format_choices + ")",
       cxxopts::value<std::string>()->default_value("flatbuffers"),
       "NAME")
      ("o,out",
       "output file name pattern",
       cxxopts::value<std::string>()->default_value((std::filesystem::current_path() / "tests" / "test_%d").string()),
       "FILE")
      ("stdout",
       "print test cases to stdout (alias for --out='')",
       cxxopts::value<bool>())
      ("n",
       "number of tests to generate",
       cxxopts::value<int>()->default_value("1"),
       "NUM")
      ("memo-size",
       "memoize the last NUM unique tests; if a memoized test case is generated again, it is discarded and generation of a unique test case is retried",
       cxxopts::value<int>()->default_value("0"),
       "NUM"
      )
      ("unique-attempts",
       "limit on how many times to try to generate a unique (i.e., non-memoized) test case; no effect if --memo-size=0",
       cxxopts::value<int>()->default_value("2"),
       "NUM"
      )
      ("random-seed",
       "initialize random number generator with fixed seed (not set by default)",
       cxxopts::value<int>(),
       "NUM")
      ("dry-run",
       "generate tests without writing them to file or printing to stdout (do not keep generated tests in population either)",
       cxxopts::value<bool>()->default_value("false"))
      ("version", "print version and exit")
      ("help", "print help and exit")
      ;
    auto args = options.parse(argc, argv);

    if (args.count("help")) {
      pout(options.help());
      exit(0);
    }
    if (args.count("version")) {
      poutf("{} {}", argv[0], GRAMMARINATOR_STRFY(GRAMMARINATOR_VERSION));
      poutf("generator: {}", GRAMMARINATOR_STRFY(GRAMMARINATOR_GENERATOR));
      poutf("model: {}", GRAMMARINATOR_STRFY(GRAMMARINATOR_MODEL));
      poutf("listener: {}", GRAMMARINATOR_STRFY(GRAMMARINATOR_LISTENER));
      poutf("transformer: {}", GRAMMARINATOR_STRFY(GRAMMARINATOR_TRANSFORMER));
      poutf("serializer: {}", GRAMMARINATOR_STRFY(GRAMMARINATOR_SERIALIZER));
      exit(0);
    }

    auto tree_format_it = tree_formats.find(args["tree-format"].as<std::string>());
    if (tree_format_it == tree_formats.end()) {
      throw cxxopts::exceptions::parsing("Invalid argument for option 'tree-format'");
    }
    std::string tree_extension = std::get<0>(tree_format_it->second);
    TreeCodec* tree_codec = std::get<1>(tree_format_it->second)();

    // Parse optional custom weights from JSON
    runtime::WeightedModel::WeightMap weights;
    if (args.count("weights")) {
      JsonWeightLoader().load(args["weights"].as<std::string>(), weights);
    }

    auto allowlist = args.count("allowlist")
        ? args["allowlist"].as<std::vector<std::string>>()
        : std::vector<std::string>{};
    for (auto& s : allowlist) {
      trim(s);
    }

    auto blocklist = args.count("blocklist")
        ? args["blocklist"].as<std::vector<std::string>>()
        : std::vector<std::string>{};
    for (auto& s : blocklist) {
      trim(s);
    }

    FilePopulation *population = args.count("population") ? new FilePopulation(args["population"].as<std::string>(), tree_extension, *tree_codec) : nullptr;
    int seed = args.count("random-seed") ? args["random-seed"].as<int>() : std::random_device()();
    GeneratorTool generator(DefaultGeneratorFactory<GRAMMARINATOR_GENERATOR, GRAMMARINATOR_MODEL, GRAMMARINATOR_LISTENER>(weights),  // generator_factory
                            args.count("stdout") ? "" : args["out"].as<std::string>(),  // out_format
                            args.count("rule") ? args["rule"].as<std::string>() : "",  // rule
                            RuleSize(args["max-depth"].as<int>(), args["max-tokens"].as<int>()),  // limit
                            population,  // population
                            args["keep-trees"].as<bool>(),  // keep_trees
                            !args["no-generate"].as<bool>(),  // generate
                            !args["no-mutate"].as<bool>(),  // mutate
                            !args["no-recombine"].as<bool>(),  // recombine
                            !args["no-grammar-violations"].as<bool>(), // unrestricted
                            std::unordered_set<std::string>(allowlist.begin(), allowlist.end()), // allowlist
                            std::unordered_set<std::string>(blocklist.begin(), blocklist.end()), // blocklist
                            GRAMMARINATOR_TRANSFORMER ? std::vector<Rule*(*)(Rule*)>{GRAMMARINATOR_TRANSFORMER} : std::vector<Rule*(*)(Rule*)>{},  // transformers
                            GRAMMARINATOR_SERIALIZER,  // serializer
                            args["memo-size"].as<int>(),  // memo-size
                            args["unique-attempts"].as<int>(),  // unique-attempts
                            args["dry-run"].as<bool>()  // dry-run
                            );

    for (int i = 0, n = args["n"].as<int>(); i < n; ++i) {
        random_engine.seed(seed + i);
        generator.create_test(i);
    }

    delete tree_codec;
    delete population;
  } catch (const cxxopts::exceptions::parsing &e) {
    perrf("error parsing options: {}", e.what());
    exit(1);
  }
}
